src/huggingface_hub/snapshot_download.py

```diff
@@ -53,18 +64,86 @@ def snapshot_download(
     else:
         token = None
 
-    _api = HfApi()
-    model_info = _api.model_info(repo_id=repo_id, revision=revision, token=token)
+    # remove all `/` occurances to correctly convert repo to directory name
+    repo_id_flattened = repo_id.replace("/", REPO_ID_SEPARATOR)
+
+    # if we have no internet connection we will look for the
+    # last modified folder in the cache
+    if local_files_only:
+        # possible repos have <path/to/cache_dir>/<flatten_repo_id> prefix
+        repo_folders_prefix = os.path.join(cache_dir, repo_id_flattened)
+
+        # list all possible folders that can correspond to the repo_id
+        # and are of the format <flattened-repo-id>.<revision>.<commit-sha>
+        # now let's list all cached repos that have to be included in the revision
+        # 1) cached repos of format <repo_id>.{revision}.<any-hash>
+        repo_folders_branch = glob(repo_folders_prefix + "." + revision + ".*")
+
+        # 2) cached repos of format <repo_id>.{revision}
+        # -> in this case {revision} has to be a commit sha
+        repo_folders_commit_only = glob(repo_folders_prefix + "." + revision)
 
-    storage_folder = os.path.join(
-        cache_dir, repo_id.replace("/", REPO_ID_SEPARATOR) + "." + model_info.sha
-    )
+        # 3) cached repos of format <repo_id>.<any-branch>.{revision}
+        # -> in this case {revision} also has to be a commit sha
+        repo_folders_branch_commit = glob(repo_folders_prefix + ".*." + revision)
```

Comment:
And this isn't necessarily a `branch` + a `commit`, I'd say it's an `explicit_revision` + a `folder_sha` (even if the latter is ultimately a commit hash)

How about `repo_folders_explicit_revision_and_sha` ?